sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -4884,6 +4967,9 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
   // in CommandData.
   PI_CALL(piKernelRetain(Kernel));
 
+  if (UseImmediateCommandLists && IndirectAccessTrackingEnabled)
+    Queue->KernelsToBeSubmitted.push_back(Kernel);
+
```

Comment:
why is this moved here?

Reply:
Because appending to the commandlist will immediately cause the kernel to start running, any book-keeping for memory tracking should be done before.

Reply:
But this will not be used until _pi_queue::executeCommandList, where handling of KernelsToBeSubmitted is. Should that be also moved here?

Reply:
Yes, the actual tracking has to be moved before the kernels starts executing. I'll carve it out into a separate function to be called from two places: a) with immediate commandlists before submission, b) when doing batching, as part of batch submission.